docs/remote/create-dev-container.md

```diff
@@ -56,253 +43,162 @@ All of the predefined container configurations you can pick from come from the [
-        "golang.go"
+        "dbaeumer.vscode-eslint"
     ],
-    "runArgs": [
-        "--cap-add=SYS_PTRACE",
-        "--security-opt",
-        "seccomp=unconfined"
-    ]
+    "forwardPorts": [ 3000 ],
+    "postCreateCommand": "sudo apt-get update && sudo apt-get install git",
 }
 ```
 
-While less efficient than a custom Dockerfile, you can also use the `postCreateCommand` property to **install additional software** that **may not be in your base image** or for cases where you would prefer **not to modify a deployment Dockerfile** you are reusing.
+The `postCreateCommand` is run once the container is running, so you can also use the property to run commands like `npm install` or to execute a shell script in your source tree (if you have mounted it).
+
+```json
+"postCreateCommand": "bash scripts/install-dev-tools.sh"
+```
+
+Rather than referencing an image directly in `devcontainer.json` or installing software via the `postCreateCommand`, an even more efficient practice is to use a Dockerfile.
 
-For example, here is a `devcontainer.json` that adds compiler tools and the [C++](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) extension to the base Ubuntu 18.04 container image:
+## Dockerfile
+
+A Dockerfile will also live in the `.devcontainer` folder. You can replace the `image` property in `devcontainer.json` with `dockerFile`:
 
 ```json
 {
-    "image": "ubuntu:18.04",
+    "dockerFile": "Dockerfile",
     "extensions": [
-        "ms-vscode.cpptools"
+        "dbaeumer.vscode-eslint"
     ],
-    "postCreateCommand": "apt-get update && apt-get install -y build-essential cmake cppcheck valgrind"
+    "forwardPorts": [ 3000 ]
 }
 ```
 
-See [installing additional software](#install-additional-software) for more information on using `apt-get` to install software.
-
-This command is run once your source code is mounted, so you can also use the property to run commands like `npm install` or to execute a shell script in your source tree.
-
-```json
-"postCreateCommand": "bash scripts/install-dev-tools.sh"
-```
+When you make changes like installing new software, changes made in the Dockerfile will persist even upon a rebuild of the dev container.
 
-By default, when VS Code starts a container, it will **override the container's default command** to be `/bin/sh -c "while sleep 1000; do :; done"`. This is done because the container will stop if the default command fails or exits. However, this may not work for certain images. If the image you are using requires the default command be run to work properly, add the following to your `devcontainer.json` file.
+In your Dockerfile, use `FROM` to designate the image, and the `RUN` [instruction](#using-an-image-or-dockerfile) to install any software. You can use `&&` to string together multiple commands.
 
-```json
-"overrideCommand": false
+```Dockerfile
+FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-12
+RUN apt-get update && apt-get install git
```

Comment:
We could show this as

```
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get install git
```

You'll see warnings if you don't... which could confuse new people as I think about it.  Not sure if the addition of the export is more confusing or not.

Reply:
I think leading users to something where they'll run into warnings and not explaining the warnings or how to combat them would be confusing. I updated `RUN` to match your suggestion and also added a note below it that explains it was added to avoid running into warnings later on.